### PR TITLE
Assignment3 @ Joshua-Pedersen

### PR DESCRIPTION
**Pull request by slanante**
Initial PR. Attempted Extra Credit portion to process a JSON format. Used the Newtonsoft.Json library. You will need to configure your VS or VSCode to handle the library or it will not compile.

### CODE REVIEW

File: CanHazFunny/CanHazFunny/JokeService.cs [hunk 1]
```diff
@@ -1,15 +1,22 @@
...
         private HttpClient HttpClient { get; } = new();
 
+        public string? Joke { get; set; }
+
         public string GetJoke()
         {
-            string joke = HttpClient.GetStringAsync("https://geek-jokes.sameerkumar.website/api").Result;
-            return joke;
+            Joke = HttpClient.GetStringAsync("https://geek-jokes.sameerkumar.website/api?format=json").Result;
+            dynamic? obj = JsonConvert.DeserializeObject(Joke);
```
Comment (jdriscoll1) [inline CanHazFunny/CanHazFunny/JokeService.cs:15]: Why are you deserializing the json, then reserializing it, then deserializing it again manually? Once it's deserialized you can parse the information, or instead manually deserialize the initially received json without deserializing it.

File: CanHazFunny/CanHazFunny/Jester.cs [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+using System.Diagnostics.Contracts;
+
+namespace CanHazFunny
+{
+    public class Jester
+    {
+        private JokeService _jokeService;
+        private JokeWriter _jokeWriter;
+        
+        public Jester(JokeService getJoke, JokeWriter writeJoke)
```
Comment (jdriscoll1) [inline CanHazFunny/CanHazFunny/Jester.cs:11]: Objects should be named as a noun, not as a verb for easier readability. A jokeGetter gets a joke sounds better than A getJoke gets a joke.

File: CanHazFunny/CanHazFunny/CanHazFunny.csproj [hunk 3]
```diff
@@ -1,8 +1,14 @@
 <Project Sdk="Microsoft.NET.Sdk">
-
   <PropertyGroup>
     <OutputType>Exe</OutputType>
-    <TargetFramework>net6.0</TargetFramework>
+    <TargetFramework>net7.0</TargetFramework>
+    <LangVersion>11.0</LangVersion>
```
Comment (jdriscoll1) [inline CanHazFunny/CanHazFunny/CanHazFunny.csproj:5]: Change to latest release

File: CanHazFunny/CanHazFunny.Tests/JesterTests.cs [hunk 4]
```diff
@@ -5,6 +5,38 @@ namespace CanHazFunny.Tests
...
+            JokeWriter testWriter = new();
+            Jester testJester = new Jester(testJoke, testWriter);
+
+            // Act
+
+            testJester.TellJoke();
+            
+            // Assert
+
+            Assert.IsFalse((testJoke.Joke!).Contains("Chuck Norris"));
```
Comment (jdriscoll1) [inline CanHazFunny/CanHazFunny.Tests/JesterTests.cs:39]: Testing one joke to see if it contains Chuck Norris is not thorough enough, either test many many jokes, or manually create a test that returns a Chuck Norris joke and validate that it does not go through.

File: CanHazFunny/CanHazFunny.Tests/JesterTests.cs [hunk 5]
```diff
@@ -5,6 +5,38 @@ namespace CanHazFunny.Tests
...
+
+            string generatedJoke = testJoke.GetJoke();
+
+            // Assert
+
+            Assert.IsNotNull(generatedJoke);
+        }
+
+        [TestMethod]
+        public void TellJoke_GenAndWrite_DoesNotContainChuck()
```
Comment (jdriscoll1) [inline CanHazFunny/CanHazFunny.Tests/JesterTests.cs:25]: Avoid abridging Generate

**Review by jdriscoll1 — COMMENTED** [review]
Create two interfaces. One to represent the JokeService and one interface to represent outputting the joke to the screen (Console.WriteLine). ✔
The JokeService will need to have the interface applied to it. ✔
Create an implementation of the output interface that writes out the joke to the console. ✔
Implement the Jester class. It should take in both interfaces as dependencies. These dependencies should be null checked. ✔
The Jester class TellJoke() method should retrieve a joke from the JokeService. If the joke contains "Chuck Norris", skip it and get another. The joke should be written to the output dependency. ✔
Unit test the Jester class. Code coverage should be above 90% for this class. Moq will make this significantly easier. ❌ 
Testing if one joke contains Chuck Norris is not thorough enough. It is better to implement another method that makes sure the joke is Chuck Norris and then makes sure it does not go through

Fundamentals
Be sure you enable:
Nullability for all projects ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ❌ 
Incorrectly set to 11.0, set to latest instead
Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
For this assignment, always use Assert.AreEqual<T>() (the generic version) ✔
Ensure there are no errors or warnings (including code analysis warnings) ✔
All of the above should be unit tested ✔
Choose simplicity over complexity ✔
Extra Credit
Unit test your implementation that writes the joke out to the screen. How hard could it be to unit test a single line method ;)? ❌ 
The Geek jokes API that is being used can also return jokes in a JSON format. Update the JokeService to retrieve jokes using JSON. GetJoke should still return a string. ✔

File: CanHazFunny/CanHazFunny/CanHazFunny.csproj (hunk 3 above)
Comment (slanante) [reply CanHazFunny/CanHazFunny/CanHazFunny.csproj:5]: Isn't net7.0 the latest release? The snippet of code attached on your review shows that net7.0 was added.

File: CanHazFunny/CanHazFunny/Jester.cs (hunk 2 above)
Comment (slanante) [reply CanHazFunny/CanHazFunny/Jester.cs:11]: Thanks! I'll go ahead and change the variable names.

File: CanHazFunny/CanHazFunny/JokeService.cs (hunk 1 above)
Comment (slanante) [reply CanHazFunny/CanHazFunny/JokeService.cs:15]: Yeah, that was messy. I rewrote this after getting a better understanding of how Json objects work.

File: CanHazFunny/CanHazFunny.Tests/JesterTests.cs (hunk 4 above)
Comment (slanante) [reply CanHazFunny/CanHazFunny.Tests/JesterTests.cs:52]: Great idea, will get that added!

File: CanHazFunny/CanHazFunny.Tests/JesterTests.cs [hunk 6]
```diff
@@ -1,10 +1,68 @@
...
+            // Assert
+
+            Assert.IsFalse((testJoke.Joke!).Contains("Chuck Norris"));
+        }
+
+        [TestMethod]
+        public void FilterJoke_If_ChuckNorris()
+        {
+            // Arrange
+            Jester jester = new Jester(_mockGetJoke!.Object, _mockWriteJoke!.Object!);
```
Comment (JLee35) [inline CanHazFunny/CanHazFunny.Tests/JesterTests.cs:59]: `Object!` unnecessary.
```suggestion
            Jester jester = new Jester(_mockGetJoke!.Object, _mockWriteJoke!.Object);
```

File: CanHazFunny/CanHazFunny.Tests/JesterTests.cs [hunk 7]
```diff
@@ -1,10 +1,68 @@
...
+using Moq;
 
 namespace CanHazFunny.Tests
 {
     [TestClass]
     public class JesterTests
     {
-        
+        private Mock<JokeService>? _mockGetJoke;
+        private Mock<JokeWriter>? _mockWriteJoke;
```
Comment (JLee35) [inline CanHazFunny/CanHazFunny.Tests/JesterTests.cs:10]:
```suggestion
        private Mock<IGetJoke>? _mockGetJoke;
        private Mock<IWriteJoke>? _mockWriteJoke;
```

File: CanHazFunny/CanHazFunny/Jester.cs [hunk 8]
```diff
@@ -0,0 +1,35 @@
+using System;
+using System.Diagnostics.Contracts;
+
+namespace CanHazFunny
+{
+    public class Jester
+    {
+        private JokeService _jokeService;
+        private JokeWriter _jokeWriter;
```
Comment (JLee35) [inline CanHazFunny/CanHazFunny/Jester.cs:9]: The point of using interfaces is to add a layer of abstraction between the depending class and the dependency. If we set the types of these two members to a concrete class, we would have to update this code if we wanted to use different classes that implemented the same interfaces.
```suggestion
        private IGetJoke_jokeService;
        private IWriteJoke_jokeWriter;
```

File: CanHazFunny/CanHazFunny/Jester.cs [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+
+        public Jester(JokeService jokeGetter, JokeWriter jokeWriter)
```
Comment (JLee35) [inline CanHazFunny/CanHazFunny/Jester.cs:11]:
```suggestion
        public Jester(IGetJoke jokeGetter, IWriteJoke jokeWriter)
```

File: CanHazFunny/CanHazFunny/JokeService.cs [hunk 10]
```diff
@@ -1,15 +1,20 @@
...
-        public string GetJoke()
+        public string? Joke { get; set; }
+
+        public virtual string GetJoke()
         {
-            string joke = HttpClient.GetStringAsync("https://geek-jokes.sameerkumar.website/api").Result;
-            return joke;
+            Joke = HttpClient.GetStringAsync("https://geek-jokes.sameerkumar.website/api?format=json").Result;
+            dynamic? obj = JsonConvert.DeserializeObject(Joke);
+            string json = obj!.joke;
```
Comment (JLee35) [inline CanHazFunny/CanHazFunny/JokeService.cs:16]: Using `dynamic?` tells the compiler that `JsonConvert.DeserializeObject(Joke)` might return a `null` reference, but then on line 16 you are referencing `obj!.joke` telling the compiler you are sure `obj` is not null. How do you know `obj` is not null if you haven't checked it? What if the user's internet connection dropped before making the API request?
```suggestion
            dynamic? obj = JsonConvert.DeserializeObject(Joke);
            
            if (obj is null) 
            {
               // Either try again, or let the user know a problem occurred.
            }
            string json = obj!.joke;
```

File: CanHazFunny/CanHazFunny/JokeWriter.cs [hunk 11]
```diff
@@ -0,0 +1,13 @@
+using System;
+
+namespace CanHazFunny
+{
+    public class JokeWriter : IWriteJoke
+    {
+
+        public virtual void SayJoke(string joke)
```
Comment (JLee35) [inline CanHazFunny/CanHazFunny/JokeWriter.cs:8]: `virtual` wouldn't be required here if you had used the interfaces when setting up your Mock.

**Comment by JLee35** [comment]
Good job on the extra credit, but the main point of using interfaces when declaring dependencies was missed. Read up on the Dependency Inversion Principal [here.](https://en.wikipedia.org/wiki/SOLID)

**Pull request closed without merging**